src/Raven.Server/ServerWide/Commands/ETL/UpdateEtlProcessStateCommand.cs

```diff
@@ -18,40 +20,78 @@ public class UpdateEtlProcessStateCommand : UpdateValueForDatabaseCommand
 
         public string NodeTag { get; set; }
 
+        public bool HasHighlyAvailableTasks;
+
         private UpdateEtlProcessStateCommand() : base(null)
         {
             // for deserialization
         }
 
         public UpdateEtlProcessStateCommand(string databaseName, string configurationName, string transformationName, long lastProcessedEtag, string changeVector,
-            string nodeTag) : base(databaseName)
+            string nodeTag, bool hasHighlyAvalilableTasks) : base(databaseName)
         {
             ConfigurationName = configurationName;
             TransformationName = transformationName;
             LastProcessedEtag = lastProcessedEtag;
             ChangeVector = changeVector;
             NodeTag = nodeTag;
+            HasHighlyAvailableTasks = hasHighlyAvalilableTasks;
         }
 
         public override string GetItemId()
         {
             return EtlProcessState.GenerateItemName(DatabaseName, ConfigurationName, TransformationName);
         }
 
+        private IDatabaseTask GetMatchingConfiguration(DatabaseRecord record)
+        {
+            var taskName = GetItemId();
+            for (var i=0; i< record.RavenEtls.Count; i++)
+            {
+                if (record.RavenEtls[i].Name == ConfigurationName)
+                {
+                    return record.RavenEtls[i];
+                }
+            }
+
+            for (var i = 0; i < record.SqlEtls.Count; i++)
+            {
+                if (record.SqlEtls[i].Name == ConfigurationName)
+                {
+                    return record.SqlEtls[i];
+                }
+            }
+
+            return null;
+        }
+
         protected override BlittableJsonReaderObject GetUpdatedValue(long index, DatabaseRecord record, JsonOperationContext context, BlittableJsonReaderObject existingValue, RachisState state)
         {
             EtlProcessState etlState;
 
             if (existingValue != null)
+            {
                 etlState = JsonDeserializationClient.EtlProcessState(existingValue);
+
+                var databaseTask = GetMatchingConfiguration(record);
+
+                if (databaseTask == null)
+                    throw new Exception($"Can't update progress of ETL {ConfigurationName} by node {NodeTag}, because it's configuration can't be found");
+
+
+                var lastResponsibleNode = GetLastResponsibleNode(HasHighlyAvailableTasks, record.Topology, NodeTag);
+                if (record.Topology.WhoseTaskIsIt(state, databaseTask, lastResponsibleNode) != NodeTag)
```

Comment:
What happens when `lastResponsibleNode` is null here?

Reply:
There's no way it's null.. it is a function, that may return null

Reply:
Oh, I missed that.